src/agent.ts

```diff
@@ -310,25 +316,140 @@ export class Agent {
+   */
+  public static async publishAgentJson(
+    connection: Connection,
+    project: SfProject,
+    agentJson: AgentJson
+  ): Promise<PublishAgent> {
+    const maybeMock = new MaybeMock(connection);
+    let developerName: string;
+
+    getLogger().debug('Publishing AfScript');
+
+    const url = '/einstein/ai-agent/v1.1/authoring/publish';
+    const response = await maybeMock.request<PublishAgentJsonResponse>('POST', url, { agentJson });
+    if (response.botId && response.botVersionId) {
+      // we've published the AgentJson, now we need to:
+      // 1. update the AuthoringBundle-meta.xml file with response.BotId
+      // 2. retrieve the new Agent metadata that's in the org
+      const defaultPackagePath = path.resolve(project.getDefaultPackage().path);
+
+      try {
+        // First update the AuthoringBundle file with the new BotId
+        // strip the "_v1" or similar from the end of a developerName, if it's present
+        developerName = agentJson.globalConfiguration.developerName.replace(/_v\d$/, '');
+        // Try to find the authoring bundle directory by recursively searching from the default package path
+        const bundleDir = findAuthoringBundle(defaultPackagePath, developerName);
+
+        if (!bundleDir) {
+          throw SfError.create({
+            name: 'Cannot Find Bundle',
+            message: `Cannot find an authoring bundle in ${defaultPackagePath} that matches ${developerName}`,
+          });
+        }
+
+        // Construct the full file path whether we found the directory or not
+        const bundleMetaPath = path.join(bundleDir, `${developerName}.authoring-bundle-meta.xml`);
+
+        const xmlParser = new XMLParser({ ignoreAttributes: false });
+        const xmlBuilder = new XMLBuilder({
+          ignoreAttributes: false,
+          format: true,
+          suppressBooleanAttributes: false,
+          suppressEmptyNode: false,
+        });
+
+        const authoringBundle = xmlParser.parse(await readFile(bundleMetaPath, 'utf-8')) as {
+          aiAuthoringBundle: { Target: string };
+        }; // all the typing we'll need
+        authoringBundle.aiAuthoringBundle.Target = developerName;
+
+        await writeFile(bundleMetaPath, xmlBuilder.build(authoringBundle));
+
+        // Now retrieve the updated agent metadata
+        const cs = await ComponentSetBuilder.build({
+          sourcepath: [path.resolve(path.join(defaultPackagePath, 'main', 'default', 'bots', developerName))],
+          org: {
+            username: connection.getUsername() as string,
+            exclude: [],
+          },
+        });
+
+        // will unset mocks so that retrieve will work
+        nock.restore();
+        nock.cleanAll();
+        nock.enableNetConnect();
+
+        const retrieve = await cs.retrieve({
+          usernameOrConnection: connection,
+          merge: true,
+          format: 'source',
+          rootTypesWithDependencies: ['Bot'],
```

Comment:
If we build the ComponentSet in the same way that `Agent.create()` does we don't need to set `rootTypesWithDependencies` because it's set for us.  We could extract that logic into a separate function and reuse it.